### PR TITLE
enhancement: add a way to register a u&p provider

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Allows adding a custom 

```js
module.exports = {
  register({ strapi }) {
    strapi
      .plugin("users-permissions")
      .service("providers-registry")
      .add("test-provider", {
        icon: "",
        enabled: true,
        grantConfig: {
          key: "",
          secret: "",
          callback: `${strapi.config.server.url}/auth/test-provider/callback`,
          scope: ["email"],
          authorize_url: "https://awesome.com/authorize",
          access_url: "https://awesome.com/token",
          oauth: 2,
        },
        async authCallback({ accessToken, providers, purest }) {
          // use whatever you want here to get the user info
          return {
            username: "test",
            email: "test",
          };
        },
      });
  },
};

```

### Why is it needed?

We cannot maintain sucha long list of providers directly in the core & need to allow users with specific usecases to create their own. We will stop adding any new provider in the core from there.

### How to test it?

- Test existing auth isn't broken
- Use the example above with an actual oauth of your choice.

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
@pwizla Just a heads up this would require some new documentation on the v5 docs but shouldn't be too difficult to document

File: packages/plugins/users-permissions/server/bootstrap/index.js [hunk 1]
```diff
@@ -9,22 +9,26 @@
...
 const _ = require('lodash');
-const urljoin = require('url-join');
 const { getService } = require('../utils');
-const getGrantConfig = require('./grant-config');
 const usersPermissionsActions = require('./users-permissions-actions');
 
 const initGrant = async (pluginStore) => {
-  const apiPrefix = strapi.config.get('api.rest.prefix');
-  const baseURL = urljoin(strapi.config.server.url, apiPrefix, 'auth');
+  const allProviders = getService('providers-registry').getAll();
```
[the remark this answers is not in the document]
Comment (derrickmehaffy) [reply packages/plugins/users-permissions/server/bootstrap/index.js:16]: In the bootstrap which was quite ugly, having them combined like this is so much nicer.

[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/plugins/users-permissions/server/bootstrap/index.js:16]: yeah the conf just moved to the providers-registry but it's exactly the same end result

**Comment by alexandrebodin** [comment]
@derrickmehaffy We have time to merge it can you see if some people that have asked for new providers or want to customize stuff could try an experimental ?

**Review by Marc-Roig — APPROVED** [review]
code looks good! And same thought as alex, if we have time I would experiment with this before merging

**Comment by alexandrebodin** [comment]
As we will release it as part of the RC we can get some testers I bet @derrickmehaffy can find some people to help :D

Let's wait for @pwizla to have time to document it and then I'll merge 👍

**Comment by alexandrebodin** [comment]
Merging, doc PR is ready

**Pull request merged**